src/search.jl

```diff
@@ -1,81 +1,119 @@
-#=
-# This is the most general (still semi-sudo-code) of the search function.
-# 
+export AbstractGibbsNetwork
 export low_energy_spectrum
+export Solution
+
+abstract type AbstractGibbsNetwork end
 
 mutable struct Solution
     energies::Vector{Float64}
     states::Vector{Vector{Int}}
     probabilities::Vector{Float64}
+    largest_discarded_probability::Float64
 end
 
-function _partition_into_unique(
-    boundary::Vector{Int}, 
-    partial_eng::Vector{T}
-    ) where {T <: Number}
+# function _partition_into_unique(
+#     boundary::Vector{Int},
+#     partial_eng::Vector{T}
+#     ) where {T <: Number}
 
-end
+# end
 
-function _merge(
-     model::Model,
-     sol::Solution,
-    )
-    boundary = zeros(Int, length(sol.states))
-    for (i, σ) ∈ enumerate(sol.states)
-        boundary[i] = generate_boundary(model, σ)
-    end
+# function _merge(
+#      network::AbstractGibbsNetwork,
+#      sol::Solution,
+#     )
+#     boundary = []
+#     for (i, σ) ∈ enumerate(sol.states)
+#         push!(boundary, generate_boundary(network, σ))
+#     end
 
-    idx = _partition_into_unique(boundary, sol.energies)
-    Solution(sol.energies[idx], sol.states[idx], sol.probabilities[idx])
-end
+#     idx = _partition_into_unique(boundary, sol.energies)
+#     Solution(
+#         sol.energies[idx],
+#         sol.states[idx],
+#         sol.probabilities[idx],
+#         sol.largest_discarded_probability)
+# end
 
-function _sort(
-    sol::Solution,
-    k::Int,
+#TODO: this can probably be done better
+function _branch_state(
+    cfg::Vector,
+    state::Vector,
+    basis::Vector,
     )
-    perm = partialsortperm(sol.probabilities, 1:k, rev=true)
-    Solution(sol.energies[perm], sol.states[perm], sol.probabilities[perm])
+    tmp = Vector{Int}[]
+    for σ ∈ basis push!(tmp, vcat(state, σ)) end
+    vcat(cfg, tmp)
 end
 
-function _update_solution!(
-    sol::Solution,
-    model::Model, 
-    )
-    prob = eng = psol = [] 
+@inline _init_solution() = (Float64[], Float64[], Vector{Int}[])
 
-    for (i, σ) ∈ enumerate(sol.states)
-        p = conditional_probability(model, σ)
-        push!(prob, sol.probabilities[i] .* p)
-        push!(eng, sol.energies[i] .+ δE(model, σ))
-        # add sol
-    end
+# TODO: logic here can probably be done better
+function _bound(pdo::Vector{Float64}, cut::Int)
```

Comment:
What does `pdo` stand for? Can we come up with a better name for this variable?

Reply:
Sure, pdo - probabilities. Done